src/tacacs_plus_fmt_plug.c

```diff
@@ -0,0 +1,347 @@
+		self->params.max_keys_per_crypt *= omp_t;
+	}
+#endif
+	saved_key = mem_calloc(sizeof(*saved_key), self->params.max_keys_per_crypt);
+	saved_len = mem_calloc(self->params.max_keys_per_crypt, sizeof(*saved_len));
+	cracked_size = sizeof(*cracked) * self->params.max_keys_per_crypt;
+	any_cracked = 0;
+	cracked = mem_calloc(cracked_size, 1);
+}
+
+static void done(void)
+{
+	MEM_FREE(saved_key);
+	MEM_FREE(saved_len);
+	MEM_FREE(cracked);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *ctcopy, *keeptr, *p;
+	int value, extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, TAG_LENGTH) != 0)
+		return 0;
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+
+	ctcopy += TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "$")) == NULL) // version / type
+		goto err;
+	if (!isdec(p))
+		goto err;
+	value = atoi(p);
+	if (value != 0)
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // pre_hash_data
+		goto err;
+	if (hexlenl(p, &extra) != 4 * 2 || extra)
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // ciphertext
+		goto err;
+	if (hexlenl(p, &extra) > MAX_CIPHERTEXT_LENGTH * 2 || extra)
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // hash_data
+		goto err;
+	if (hexlenl(p, &extra) != 2 * 2 || extra)
+		goto err;
+
+
+	MEM_FREE(keeptr);
+	return 1;
+
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs;
+	char *ctcopy = strdup(ciphertext);
+	char *keeptr = ctcopy;
+	char *p;
+	int i;
+
+	memset(&cs, 0, SALT_SIZE);
+	ctcopy += TAG_LENGTH;
+	p = strtokm(ctcopy, "$");
+	p = strtokm(NULL, "$");
+	cs.pre_hash_data_len = strlen(p) / 2;
+	for (i = 0; i < cs.pre_hash_data_len; i++)
+		cs.pre_hash_data[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	p = strtokm(NULL, "$");
+	cs.ctlen = strlen(p) / 2;
+	for (i = 0; i < cs.ctlen; i++)
+		cs.ciphertext.buf[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	p = strtokm(NULL, "$");
+	cs.hash_data_len = 2;
+	for (i = 0; i < 2; i++)
+		cs.hash_data[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	MD5_Init(&cs.pctx);
+	MD5_Update(&cs.pctx, cs.pre_hash_data, cs.pre_hash_data_len);
+
+	MEM_FREE(keeptr);
+
+	return &cs;
+}
+
+static int check_password(int index, struct custom_salt *cs)
+{
+	MD5_CTX cur;
+	union {
+		uint64_t chunks[MD5_DIGEST_LENGTH / sizeof(uint64_t)];
+		unsigned char buf[MD5_DIGEST_LENGTH];
+	} digest;
+	union {
+		uint64_t chunks[MD5_DIGEST_LENGTH / sizeof(uint64_t)];
+		unsigned char buf[MD5_DIGEST_LENGTH];
+	} cleartext;
+	unsigned char status, flags;
+	unsigned short server_msg_len, data_len;
+
+	memcpy((void *)&cur, &cur_salt->pctx, sizeof(MD5_CTX));
+	MD5_Update(&cur, saved_key[index], saved_len[index]);
+	MD5_Update(&cur, cs->hash_data, cs->hash_data_len);
+	MD5_Final(digest.buf, &cur);
+
+	// XOR the first 8 bytes of "ciphertext" with "digest"
+	cleartext.chunks[0] = cs->ciphertext.chunks[0] ^ digest.chunks[0];
```

Comment:
Do we really need the `cleartext` variable? We could instead `digest.chunks[0] ^= ...` and use `digest` in the following code.